### PR TITLE
added support for Azure arc

### PR DESCRIPTION
**Pull request by 4gust**
# Acquire Token for Azure Arc 

## Description

This PR adds functionality for acquiring tokens for Azure Arc.
This includes the following:

- Acquiring token via SAMI
- Error handling for UAMI
- Environment detection
- Probing

## Changes

- Some general refactoring of names etc
- Adding required variables for Azure Arc
- Source retrieval
- Retrieval of endpoints via env vars
- Validation logic for env vars and checking for himds
- Handling of Azure Arc request, along with the mandatory 401 return and retry
- Some changes to tests

## Testing

- Added unit tests to verify logic for `GetSource`, `getAzureArcEnvironmentVariables`, `validateAzureArcEnvironment`, `fileExists` and `handleAzureArcResponse`
- Updates tests for `AcquireToken` to check for AzureArc

### CODE REVIEW

File: apps/managedidentity/managedidentity.go [hunk 1]
```diff
@@ -135,65 +194,171 @@ func New(id ID, options ...ClientOption) (Client, error) {
...
+	imdsEndpoint := os.Getenv(imdsEndVar)
+
+	if identityEndpoint != "" && identityHeader != "" {
+		if identityServerThumbprint != "" {
+			return ServiceFabric, nil
+		}
+		return AppService, nil
+	} else if msiEndpoint != "" {
+		return CloudShell, nil
+	} else if isAzureArcEnvironment(identityEndpoint, imdsEndpoint, runtime.GOOS) {
```
Comment (bgavrilMS) [inline apps/managedidentity/managedidentity.go:227]: Considre that you move the runtim.GOOS inside `isAzureArcEnvironment`.

File: apps/managedidentity/managedidentity.go [hunk 2]
```diff
@@ -135,65 +194,171 @@ func New(id ID, options ...ClientOption) (Client, error) {
...
-		msiParameters.Set(miQueryParameterClientId, string(t))
-	case UserAssignedResourceID:
-		msiParameters.Set(miQueryParameterResourceId, string(t))
-	case UserAssignedObjectID:
-		msiParameters.Set(miQueryParameterObjectId, string(t))
-	case systemAssignedValue: // not adding anything
+	// ignore cached access tokens when given claims
+	if o.claims == "" {
+		if cacheManager == nil {
+			return base.AuthResult{}, errors.New("cache instance is nil")
```
Comment (bgavrilMS) [inline apps/managedidentity/managedidentity.go:247]: This error can never occur? App developer doesn't configure the cache for MSI. Why are we throwing?

File: apps/managedidentity/managedidentity.go [hunk 3]
```diff
@@ -135,65 +194,171 @@ func New(id ID, options ...ClientOption) (Client, error) {
...
-	switch t := id.(type) {
-	case UserAssignedClientID:
-		msiParameters.Set(miQueryParameterClientId, string(t))
-	case UserAssignedResourceID:
-		msiParameters.Set(miQueryParameterResourceId, string(t))
-	case UserAssignedObjectID:
-		msiParameters.Set(miQueryParameterObjectId, string(t))
-	case systemAssignedValue: // not adding anything
+	// ignore cached access tokens when given claims
+	if o.claims == "" {
```
Comment (bgavrilMS) [inline apps/managedidentity/managedidentity.go:245]: Do we also have `force refresh` ?

[the remark this answers is not in the document]
Comment (bgavrilMS) [reply apps/managedidentity/managedidentity.go:245]: Do we have any tests for this?

[the remark this answers is not in the document]
Comment (4gust) [reply apps/managedidentity/managedidentity.go:245]: in Go string are a "" base value
so it will always be "" and not nil as we are creating a struct.

Comment (4gust) [reply apps/managedidentity/managedidentity.go:245]: We did not add that, should we add that ?

File: apps/managedidentity/managedidentity.go (hunk 2 above)
Comment (4gust) [reply apps/managedidentity/managedidentity.go:247]: removed

File: apps/managedidentity/managedidentity_test.go [hunk 4]
```diff
@@ -217,6 +263,271 @@ func Test_SystemAssigned_Returns_Token_Success(t *testing.T) {
...
+			if err != nil {
+				t.Fatalf(errorFormingJsonResponse, err.Error())
+			}
+
+			mockClient.AppendResponse(mock.WithHTTPStatusCode(http.StatusOK), mock.WithHTTPHeader(headers),
+				mock.WithBody(responseBody))
+
+			client, err := New(SystemAssigned(), WithHTTPClient(&mockClient))
+			if err != nil {
+				t.Fatalf(`expected error: "%v" "`, testCase.expectedError)
```
Comment (chlowell) [inline apps/managedidentity/managedidentity_test.go:464]: 1. when failing due to an error, include that error
2. you expect an error from `AcquireToken`, not `New`
```suggestion
				t.Fatal(err)
```

File: apps/managedidentity/managedidentity_test.go [hunk 5]
```diff
@@ -217,6 +263,271 @@ func Test_SystemAssigned_Returns_Token_Success(t *testing.T) {
...
+				return
+			}
+			result, err := client.AcquireToken(context.Background(), resource)
+			if err == nil || !strings.Contains(err.Error(), testCase.expectedError) {
+				t.Fatalf(`expected error: "%v" got error: "%v"`, testCase.expectedError, err)
+
+			}
+			if result.AccessToken != "" {
+				t.Fatalf("access token should be empty")
```
Comment (chlowell) [inline apps/managedidentity/managedidentity_test.go:473]:
```suggestion
				t.Fatal("access token should be empty")
```

File: apps/managedidentity/managedidentity.go [hunk 6]
```diff
@@ -135,65 +194,168 @@ func New(id ID, options ...ClientOption) (Client, error) {
...
 	return client, nil
 }
 
-func createIMDSAuthRequest(ctx context.Context, id ID, resource string, claims string) (*http.Request, error) {
-	var msiEndpoint *url.URL
-	msiEndpoint, err := url.Parse(imdsEndpoint)
-	if err != nil {
-		return nil, fmt.Errorf("couldn't parse %q: %s", imdsEndpoint, err)
+// GetSource detects and returns the managed identity source available on the environment.
+func getSource() (Source, error) {
```
[the remark this answers is not in the document]
Comment (4gust) [reply apps/managedidentity/managedidentity.go:213]: I assumed we are not going to use this outside, I will expose this

**Review by bgavrilMS — APPROVED** [review]

File: apps/managedidentity/managedidentity_test.go [hunk 7]
```diff
@@ -26,14 +26,14 @@ const (
...
-	wrongSecretKey        = "2secret.key"
 	basicRealm            = "Basic realm="
-	thisShouldFail        = "This should fail"
 
 	errorExpectedButGot      = "expected %v, got %v"
 	errorFormingJsonResponse = "error while forming json response : %s"
 )
 
+// We only need to do this if we'll routinely run the test suite on Arc-managed machines (I doubt we will). If you still want to keep this, it would be best to use a path that's certain not to exist such as filepath.Join(t.TempDir(), t.Name())
+
```
Comment (chlowell) [inline apps/managedidentity/managedidentity_test.go:36]: 👀

File: apps/managedidentity/managedidentity.go [hunk 8]
```diff
@@ -202,74 +349,137 @@ func (client Client) getTokenForRequest(req *http.Request) (accesstokens.TokenRe
...
+	msiEndpoint, parseErr := url.Parse(identityEndpoint)
+
+	if parseErr != nil {
+		return nil, fmt.Errorf("couldn't parse %q: %s", identityEndpoint, parseErr)
 	}
-	tokenResponse, err := client.getTokenForRequest(req)
+
+	msiParameters := msiEndpoint.Query()
+	msiParameters.Set(apiVersionQueryParameterName, azureArcAPIVersion)
+	resource = strings.TrimSuffix(resource, "/.default")
```
Comment (chlowell) [inline apps/managedidentity/managedidentity.go:407]: AcquireToken does this now
```suggestion
```

**Review by chlowell — APPROVED** [review]

File: apps/managedidentity/managedidentity.go [hunk 9]
```diff
@@ -35,23 +38,64 @@ const (
...
 	systemAssignedManagedIdentity = "system_assigned_managed_identity"
+
+	// Azure Arc
+	azureArcEndpoint               = "http://127.0.0.1:40342/metadata/identity/oauth2/token"
+	azureArcAPIVersion             = "2020-06-01"
+	azureArcFileExtension          = ".key"
+	azureArcMaxFileSizeBytes int64 = 4096
+	linuxTokenPath                 = "/var/opt/azcmagent/tokens"
+	linuxHimdsPath                 = "/opt/azcmagent/bin/himds"
+	azureConnectedMachine          = "AzureConnectedMachine"
```
Comment (neha-bhargava) [inline apps/managedidentity/managedidentity.go:64]:
```suggestion
	azureConnectedMachine          = "AzureConnectedMachineAgent"
```

Comment (neha-bhargava) [reply apps/managedidentity/managedidentity.go:64]: Have you tested the flow manually? If not I would recommend to test it before merging.

File: apps/managedidentity/managedidentity.go [hunk 10]
```diff
@@ -35,23 +38,64 @@ const (
...
+	// Environment Variables
+	identityEndpointEnvVar              = "IDENTITY_ENDPOINT"
+	identityHeaderEnvVar                = "IDENTITY_HEADER"
+	azurePodIdentityAuthorityHostEnvVar = "AZURE_POD_IDENTITY_AUTHORITY_HOST"
+	imdsEndVar                          = "IMDS_ENDPOINT"
+	msiEndpointEnvVar                   = "MSI_ENDPOINT"
+	identityServerThumbprintEnvVar      = "IDENTITY_SERVER_THUMBPRINT"
 )
 
+var getAzureArcPlatformPath = func(platform string) string {
```
Comment (neha-bhargava) [inline apps/managedidentity/managedidentity.go:77]:
```suggestion
var getAzureArcTokenPath = func(platform string) string {
```

Comment (neha-bhargava) [reply apps/managedidentity/managedidentity.go:77]: Nit: rename the variable name. It is a bit confusing.

File: apps/managedidentity/managedidentity.go [hunk 11]
```diff
@@ -35,23 +38,64 @@ const (
...
+	case "windows":
+		return filepath.Join(os.Getenv("ProgramData"), azureConnectedMachine, tokenName)
+	case "linux":
+		return linuxTokenPath
+	default:
+		return ""
+	}
+}
+
+var getAzureArcFilePath = func(platform string) string {
```
Comment (neha-bhargava) [inline apps/managedidentity/managedidentity.go:88]:
```suggestion
var getAzureArcHimdsFilePath = func(platform string) string {
```

**Review by neha-bhargava — CHANGES_REQUESTED** [review]
The azure arc file path is not correct for windows. I would recommend to perform a manual test before completing this PR.

**Pull request merged**